### ISSUE TITLE
Can't Read Flac Tags

### ISSUE THREAD

**Issue by astralspurs** [Martchus/tageditor#36]
I can't find anything in the help files related specifically to editing FLAC files so I'm not sure if I'm missing an option I should use for FLAC containers. I'm trying to read and write seemingly normal FLAC files.

```
- Container format: raw Free Lossless Audio Codec frames
 - Tracks: FLAC-2ch
    ID                            0
    Type                          Audio
    Format                        Free Lossless Audio Codec
    Abbreviation                  FLAC
    Duration                      4 min 49 s 45 µs 300 ns
    Channel count                 2
    Bits per sample               16
    Sampling frequency            44100 Hz
    Sample count                  12744902
```
When I run 'tageditor -i -v -f ./*.flac' I get the following diagnostic messages:
```
Diagnostic messages:
    Information  05:11:46   parsing chapters: Parsing chapters is not implemented for the container format of the file.
    Information  05:11:46   parsing attachments: Parsing attachments is not implemented for the container format of the file.
```

When I try to set with the verbose option, I don't get any messages reported to me.
```
Setting tag information for "./09 Adapting to Life in a Tiny Pea Chute.flac" ...
 - [100%] Updating ID3v2 tags ...
 - [100%] Preparing streams for rewriting ...
 - [100%] Writing frames ...
 - Changes have been applied.
 - Diagnostic messages:
    Information  05:24:53   making file: Changes are about to be applied.
```
Getting tags info just gives me:

`- File has no (supported) tag information.`

Cheers

**Comment by Martchus** [comment]
> I can't find anything in the help files related specifically to editing FLAC files so I'm not sure if I'm missing an option I should use for FLAC containers.

There's nothing special to take into account when dealing with FLAC files.

> When I try to set with the verbose option, I don't get any messages reported to me.

Though, which command did you use exactly to set the tag info?

That the messages regarding chapters and attachments are only printed with `-i` option is because only when showing technical info it attempts to parse those.

> Getting tags info just gives me:
> `- File has no (supported) tag information:`

And is this at least correct?

I actually [have tests](https://github.com/Martchus/tagparser/blob/master/tests/overallflac.cpp) for FLAC. But I have to admit that those tests are not very extensive and to not tests with a variety of different FLAC files. Can you share the failing file? This would help me fixing this because I need to be able to reproduce locally to see what's going wrong.

**Comment by Martchus** [comment]
I did some tests with my own files and could find and fix the following FLAC-related issues:

1. Incorrect handling of "isLast" flag. This might have prevented adding a new tag correctly, when no tag was present before.
2. Showing tag information for FLAC files in CLI didn't work.
3. Removing FLAC tags in GUI didn't work.

I think this issue was about 1. and 2.  
Can you test whether it works with my latest changes or provide your file so I can test?

Note that all [relevant changes have been made in the underlying library](https://github.com/Martchus/tagparser/compare/v7.0.0...9cbc708a56f912e5573f8ec242adf38cc9c41d5e).

**Comment by Martchus** [comment]
> Pardon my ignorance, but I downloaded tageditor from the AUR. Do I need to build from source to get your latest changes?

Installing from the AUR means downloading from source (except you used my binary repository). I also provide [PKGBUILDs](https://github.com/Martchus/PKGBUILDs) for the Git version. Note that you should then build the Git version of `c++utilities`, `tagparser`, `qtutilities` and `tageditor`. Or just follow the steps under "Building this straight" in the README if you don't want to create packages for this.

> A music player can see the tags but tageditor cannot.

Yes, and I think I have fixed this now with my latest changes. Note that only the CLI was affected. As an optimization I prevented parsing the tracks here. But for FLAC it is still required because it stores the meta-data on track level.

**Comment by astralspurs** [comment]
So, testing it more here, I mispoke earlier. Writing is not a problem (even with the previous version I was using, 3.0.0), it's just the getting the tag information that is the problem. I can use tageditor to change tags, but to check the change I have to use another music player currently.

I built a new version and I continue to not be able to read the tag information.

Here's the output of the top of the `tageditor --help`. Did I get the proper versions of everything?

```
Tag Editor, version 3.0.0-392.3411ee3
Linked against: C++ Utilities: 4.13.0-37.c2251d5, Qt Utilities: 5.9.0, Tag Parser: 7.0.0, Qt Qml: 5.10.1, Qt WebKitWidgets: 5.212.0, Qt Concurrent: 5.10.1, Qt Network: 5.10.1, Qt Widgets: 5.10.1, Qt Core: 5.10.1
```

**Comment by Martchus** [comment]
No, you likely didn't rebuild tagparser. Can't tell for sure though, because --help shows versions present at build time and not versions present at runtime.

**Comment by astralspurs** [comment]
Darn. I followed the Building It Straight guide, so I cloned the folders c++utilities, tagparser, qtutilities and tageditor as you recommended and then ran cmake. Not sure how else to get a more recent of tagparser or test what version I currently have.

**Comment by Martchus** [comment]
> Is there a way to test what version I currently have of tagparser?

If you follow "Building it Straight", you will end up with the latest development version. You can use the usual Git commands to find out the latest commit for the different repositories.

For the packaged version you can obviously just use pacman.

> Not sure how else to get a more recent of tagparser.

Like I said before, I provide PKGBUILDs for the Git version. Since those are not in the AUR, you would need to checkout [PKGBUILDs](https://github.com/Martchus/PKGBUILDs) manually and invoke `makepkg` manually.

> I followed the Building It Straight guide, so I cloned the folders c++utilities, tagparser, qtutilities and tageditor as you recommended and then ran cmake.

And, does it work now? You can test the version you've just built by simply executing the `tageditor` binary in the built directory. There's no need to install anything. CMake will also ensure, that the libraries from the custom build are used (rather than the libs you might have still installed from the AUR packages).

**Comment by astralspurs** [comment]
Alrighty, a lot of this was a first for me so thanks for bearing with me. When I built from git before, I guess tageditor was still using the older tagparser I had installed. I proved this by uninstalling tageditor, it's dependencies and tagparser, which left my new build of tageditor broken missing the tagparser library.

I reinstalled from your PKGBUILDs and everything is working now. Thanks very much. :sweat_smile:
